block-layer/codecs/dag-jose.md

```diff
@@ -0,0 +1,100 @@
+
+### Mapping from the JOSE general JSON serialization to dag-jose serialization
+
+Both JWS and JWE supports three different serialization formats: `Compact Serialization`, `Flattened JSON Serialization`, and `General JSON Serialization`. The first two are more concise, but they only allow for one recipient. Therefore DAG JOSE always uses the `General Serialization` which ensures maximum compatibility with minimum ambiguity. Libraries implementing serialization should accept all JOSE formats including the `Decoded Representation` (see below) and convert them if necessary.
+
+To map the general JSON serialization to CBOR we do the following:
+
+- Any field which is represented as `base64url(<data>)` we map directly to `Bytes` . For fields like `header` and `protected` which are specified as the `base64url(ascii(<some json>))` that means that the value is the `ascii(<some json>)` bytes.
+- For JWS we specify that the `payload` property MUST be a CID, and we set the `payload` of the encoded JOSE object to `Bytes` containing the bytes of the CID. For applications where an additional network request to retrieve the linked content is undesirable then an `identity` multihash should be used.
+- For JWE objects the `ciphertext` must decrypt to a plaintext which is the bytes of a CID. This is for the same reason as the `payload` being a CID, and the same approach of using an `identity` multihash can be used, and most likely will be the only way to retain the confidentiality of data.
+
+Below we present an IPLD schema representing the encoded JOSE objects. Note that the `EncodedJOSE` union is not in fact a valid IPLD schema as there is no valid discriminator. The actual wire format is a single struct which contains all the keys from both the `EncodedJWE` and the `EncodedJWS` structs, implementors should follow [section 9 of the JWE spec](https://tools.ietf.org/html/rfc7516#section-9) and distinguish between these two branches of the union by checking if the `payload` attribute exists, and hence you have a JWS; or the `ciphertext` attribute, hence you have a JWE.
+
+**Encoded JOSE**
+
+```ipldsch
+type EncodedSignature struct {
+  header optional {String:Any}
+  protected optional Bytes
+  signature Bytes
+}
+
+type EncodedRecipient struct {
+  encrypted_key optional Bytes
+  header optional {String:Any}
+}
+
+type EncodedJWE struct {
+  aad optional Bytes
+  ciphertext Bytes
+  iv optional Bytes
+  protected optional Bytes
+  recipients [EncodedRecipient]
+  tag optional Bytes
+  unprotected optional {String:Any}
+}
+
+type EncodedJWS struct {
+  signatures [EncodedSignature]
+  payload optional Bytes
+}
+
+type EncodedJOSE union { EncodedJWE | EncodedJWS }
```

Comment:
We've still got to do something about this "union".  The problem discussed in [the earlier comment thread](https://github.com/ipld/specs/pull/269#discussion_r448260671) still exists.

1. This doesn't "compile".  A union type in an IPLD Schema *must* have a representation strategy.  
2. _This doesn't give me instructions on how to figure out which of these I'm dealing with._  (Which is the reason that the above is true.)

The former we could shrug off and just say "this is for approximate documentation purposes and doesn't compile"... but the second keeps me up either way.

I think my proposal is actually that we just drop this type from the schema part of the document, and do it in prose instead... along with prose instructions for how to actually distinguish these.

Ideally it would be nice to have an explicit decision tree, like: "EncodedJWE is distinguished when the first key of the top level map is 'aad'; EncodedJWS is distinguished when the first key of the top level map is 'signatures'; any other first key is a parse error."  Stating the rejection cases is something we've seen to be important to cover up front, or incompatibilities are very likely to breed in the grey areas.

Reply:
Thanks, your suggestion makes sense and I actually thought we had updated the doc to say this.. Will make sure to fix.